### PR TITLE
React to ANCM OnAsyncCompletion changes

### PR DESCRIPTION
**Pull request by jkotalik**
Fix for #442. Associated issue in ANCM: https://github.com/aspnet/AspNetCoreModule/pull/184
Still need to enable websockets, working on that currently.

### CODE REVIEW

File: samples/NativeIISSample/Startup.cs [hunk 1]
```diff
@@ -64,10 +86,64 @@ public static void Main(string[] args)
         {
             var host = new WebHostBuilder()
                 .UseIISIntegration()
+                .UseContentRoot(Directory.GetCurrentDirectory())
```
Comment (davidfowl) [inline samples/NativeIISSample/Startup.cs]: Why?

File: src/Microsoft.AspNetCore.Server.IISIntegration/NativeMethods.cs [hunk 2]
```diff
@@ -62,18 +63,33 @@ public enum REQUEST_NOTIFICATION_STATUS
...
         public unsafe static extern bool http_shutdown();
 
+        [DllImport(AspNetCoreModuleDll)]
+        public unsafe static extern int http_websockets_read_bytes(IntPtr pHttpContext, byte* pvBuffer, int cbBuffer, PFN_ASYNC_COMPLETION pfnCompletionCallback, IntPtr pvCompletionContext, out int dwBytesReceived, out bool fCompletionExpected);
+
+        [DllImport(AspNetCoreModuleDll)]
+        internal unsafe static extern int http_websockets_write_bytes(IntPtr pHttpContext, HttpApiTypes.HTTP_DATA_CHUNK* pDataChunks, int nChunks, PFN_ASYNC_COMPLETION pfnCompletionCallback, IntPtr pvCompletionContext, out bool fCompletionExpected);
+
+        [DllImport(AspNetCoreModuleDll)]
+        public unsafe static extern int http_websockets_flush_bytes(IntPtr pHttpContext, PFN_ASYNC_COMPLETION pfnCompletionCallback, IntPtr pvCompletionContext, out bool fCompletionExpected);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/NativeMethods.cs]: This isn't required.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.FeatureCollection.cs [hunk 3]
```diff
@@ -266,11 +266,11 @@ void IHttpResponseFeature.OnCompleted(Func<object, Task> callback, object state)
                 throw new InvalidOperationException("CoreStrings.UpgradeCannotBeCalledMultipleTimes");
             }
 
-            _wasUpgraded = true;
-
             StatusCode = StatusCodes.Status101SwitchingProtocols;
             ReasonPhrase = ReasonPhrases.GetReasonPhrase(StatusCodes.Status101SwitchingProtocols);
+            NativeMethods.http_enable_websockets(_pHttpContext);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.FeatureCollection.cs]: Do it after the upgrade completes?

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs [hunk 4]
```diff
@@ -42,9 +43,11 @@ internal abstract partial class HttpProtocol : NativeRequestContext, IDisposable
 
         private GCHandle _thisHandle;
         private BufferHandle _inputHandle;
-        private IISAwaitable _readOperation = new IISAwaitable();
-        private IISAwaitable _writeOperation = new IISAwaitable();
-        private IISAwaitable _flushOperation = new IISAwaitable();
+        private IISAwaitable _operation = new IISAwaitable();
+
+        private IISAwaitable _readWebSocketsOperation = new IISAwaitable();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs]: Only allocate these when the upgrade happens.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs [hunk 5]
```diff
@@ -162,6 +169,14 @@ internal unsafe HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
         private HeaderCollection HttpResponseHeaders { get; set; }
         internal HttpApiTypes.HTTP_VERB KnownMethod { get; }
 
+        private enum CurrentOperationType
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs]: Move private nested types to the bottom of the class

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs [hunk 6]
```diff
@@ -193,14 +208,25 @@ private IISAwaitable DoFlushAsync()
...
+                        CompleteFlushWebSockets(hr, 0);
+                    }
+                    return _flushWebSocketsOperation;
+                }
+                else
+                {
+                    hr = NativeMethods.http_flush_response_bytes(_pHttpContext, out var fCompletionExpected);
+                    if (!fCompletionExpected)
+                    {
+                        OnAsyncCompletion(hr, 0);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs]: So this is code reuse but you know what the operation is so you can just call complete on the `_operation directly.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs [hunk 7]
```diff
@@ -193,14 +208,25 @@ private IISAwaitable DoFlushAsync()
             unsafe
             {
                 var hr = 0;
-                hr = NativeMethods.http_flush_response_bytes(_pHttpContext, IISAwaitable.FlushCallback, (IntPtr)_thisHandle, out var fCompletionExpected);
-
-                if (!fCompletionExpected)
+                if (_wasUpgraded)
                 {
-                    CompleteFlush(hr, 0);
+                    hr = NativeMethods.http_websockets_flush_bytes(_pHttpContext, IISAwaitable.FlushCallback, (IntPtr)_thisHandle, out var fCompletionExpected);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs]: When does this call happen? Once as part of upgrade?

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs [hunk 8]
```diff
@@ -530,7 +556,21 @@ public void StartReadingRequestBody()
...
                     {
-                        int read = await ReadAsync(wb.Buffer.Length);
+                        int read = 0;
+                        if (_wasUpgraded)
+                        {
+                            await ReadAsync(wb.Buffer.Length);
+                        }
+                        else
+                        {
+                            currentOperation = currentOperation.ContinueWith(async (t) =>
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs]: We need to remove the closure allocations here. ContinueWith takes a state object. You may need to store the amount read and the length in a field. This is what the pattern looks like:

```C#
private static Func<Task, object, Task> _doReadAsync = DoReadAsync;
private int _read;
private int _length;

private static Task DoReadAsync(Task previous, object state)
{
    return ((HttpProtocol)state).DoReadAsync();
}

private async Task DoReadAsync()
{
    _currentAsyncOperation = CurrentOperationType.Read;
    _read = await ReadAsync(wb.Buffer.Length);
}


currentOperation = currentOperation.ContinueWith(_doReadAsync, this).Unwrap();

await currentOperation;
```

We might also want to come up with something less allocatey once we get it all working.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 9]
```diff
@@ -90,10 +91,18 @@ private static bool HandleShutdown(IntPtr pvRequestContext)
...
+            context.OnAsyncCompletion(hr, bytes);
+            return NativeMethods.REQUEST_NOTIFICATION_STATUS.RQ_NOTIFICATION_PENDING; // TODO
+        }
+
         private static void CompleteRequest(HttpProtocol context)
         {
             // Post completion after completing the request to resume the state machine
-            context.PostCompletion();
+            //context.PostCompletion();
+            context.IndicateCompletion(NativeMethods.REQUEST_NOTIFICATION_STATUS.RQ_NOTIFICATION_CONTINUE);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: Why?

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 10]
```diff
@@ -90,10 +91,18 @@ private static bool HandleShutdown(IntPtr pvRequestContext)
             return true;
         }
 
+        private static NativeMethods.REQUEST_NOTIFICATION_STATUS OnAsyncCompletion(IntPtr pvManagedHttpContext, int hr, int bytes)
+        {
+            var context = (HttpProtocol)GCHandle.FromIntPtr(pvManagedHttpContext).Target;
+            context.OnAsyncCompletion(hr, bytes);
+            return NativeMethods.REQUEST_NOTIFICATION_STATUS.RQ_NOTIFICATION_PENDING; // TODO
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: TODO what?

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs (hunk 7 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs]: It shouldn't be called at all. Will remove.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs (hunk 9 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: I should talk to @pan-wang about this. Post completion was not causing Continue to be called.

**Comment by jkotalik** [comment]
@davidfowl 🆙 📅

File: samples/NativeIISSample/Startup.cs [hunk 11]
```diff
@@ -1,19 +1,41 @@
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (jkotalik) [inline samples/NativeIISSample/Startup.cs]: This is just web sockets sample stuff. Should I delete it?

File: samples/NativeIISSample/wwwroot/ws.html [hunk 12]
```diff
@@ -0,0 +1,42 @@
+<!DOCTYPE html>
```
Comment (jkotalik) [inline samples/NativeIISSample/wwwroot/ws.html]: More web socket sample stuff.

File: src/Microsoft.AspNetCore.Server.IISIntegration/NativeMethods.cs [hunk 13]
```diff
@@ -62,18 +63,33 @@ public enum REQUEST_NOTIFICATION_STATUS
...
+        public unsafe static extern int http_websockets_read_bytes(IntPtr pHttpContext, byte* pvBuffer, int cbBuffer, PFN_ASYNC_COMPLETION pfnCompletionCallback, IntPtr pvCompletionContext, out int dwBytesReceived, out bool fCompletionExpected);
+
+        [DllImport(AspNetCoreModuleDll)]
+        internal unsafe static extern int http_websockets_write_bytes(IntPtr pHttpContext, HttpApiTypes.HTTP_DATA_CHUNK* pDataChunks, int nChunks, PFN_ASYNC_COMPLETION pfnCompletionCallback, IntPtr pvCompletionContext, out bool fCompletionExpected);
+
+        [DllImport(AspNetCoreModuleDll)]
+        public unsafe static extern int http_enable_websockets(IntPtr pHttpContext);
+
+        [DllImport(AspNetCoreModuleDll)]
+        public unsafe static extern int http_cancel_io(IntPtr pHttpContext);
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IISIntegration/NativeMethods.cs]: Will be used for canceling pending async reads/writes.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs [hunk 14]
```diff
@@ -53,14 +56,18 @@ internal abstract partial class HttpProtocol : NativeRequestContext, IDisposable
...
         internal unsafe HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
-            : base((HttpApiTypes.HTTP_REQUEST*) NativeMethods.http_get_raw_request(pHttpContext))
+            : base((HttpApiTypes.HTTP_REQUEST*)NativeMethods.http_get_raw_request(pHttpContext))
         {
             _thisHandle = GCHandle.Alloc(this);
 
             _pipeFactory = pipeFactory;
             _pHttpContext = pHttpContext;
 
+            NativeMethods.http_set_managed_context(pHttpContext, (IntPtr)_thisHandle);
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs]: Creates the stored context in native. Native will use it for on async completion (and eventually abort).

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs [hunk 15]
```diff
@@ -219,6 +227,35 @@ private IISAwaitable DoFlushAsync()
             }
         }
 
+        public async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs]: Fix for #435

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs (hunk 9 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: We should discuss this. I think we do want to continue on the IIS thread pool not on the .NET Thread pool.

**Comment by jkotalik** [comment]
🏓

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs (hunk 9 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: @pan-wang agreed that we should be calling indicate completion here.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs (hunk 8 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs]: #451

**Comment by jkotalik** [comment]
@davidfowl the native stuff is merged in.

File: src/Microsoft.AspNetCore.Server.IISIntegration/NativeMethods.cs [hunk 16]
```diff
@@ -34,6 +34,7 @@ public enum REQUEST_NOTIFICATION_STATUS
 
         public delegate REQUEST_NOTIFICATION_STATUS PFN_REQUEST_HANDLER(IntPtr pHttpContext, IntPtr pvRequestContext);
         public delegate bool PFN_SHUTDOWN_HANDLER(IntPtr pvRequestContext);
+        public delegate REQUEST_NOTIFICATION_STATUS PFN_MANAGED_CONTEXT_HANDLER(IntPtr pvManagedHttpContext, int hr, int bytes);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/NativeMethods.cs]: This name should be `PFN_ASYNC_COMPLETION`

File: src/Microsoft.AspNetCore.Server.IISIntegration/NativeMethods.cs (hunk 13 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Server.IISIntegration/NativeMethods.cs]: Not sure we need this but I guess it doesn't hurt.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.FeatureCollection.cs [hunk 17]
```diff
@@ -16,7 +16,7 @@
 
 namespace Microsoft.AspNetCore.Server.IISIntegration
 {
-    internal partial class HttpProtocol : IFeatureCollection,
+    internal partial class IISHttpContext : IFeatureCollection,
                                           IHttpRequestFeature,
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.FeatureCollection.cs]: nit: fix formatting

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs [hunk 18]
```diff
@@ -193,14 +200,15 @@ private IISAwaitable DoFlushAsync()
             unsafe
             {
                 var hr = 0;
-                hr = NativeMethods.http_flush_response_bytes(_pHttpContext, IISAwaitable.FlushCallback, (IntPtr)_thisHandle, out var fCompletionExpected);
-
+                hr = NativeMethods.http_flush_response_bytes(_pHttpContext, out var fCompletionExpected);
                 if (!fCompletionExpected)
                 {
-                    CompleteFlush(hr, 0);
+                    FreePinnedHeaders(_pinnedHeaders);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs]: Does this get freed in 2 places?

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs [hunk 19]
```diff
@@ -824,22 +927,30 @@ public void IndicateCompletion(NativeMethods.REQUEST_NOTIFICATION_STATUS notific
...
         {
-            _writeOperation.Complete(hr, cbBytes);
+            switch (_currentOperationType)
+            {
+                case CurrentOperationType.Read:
+                case CurrentOperationType.Write:
+                    _operation.Complete(hr, cbBytes);
+                    break;
+                case CurrentOperationType.Flush:
+                    FreePinnedHeaders(_pinnedHeaders);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs]: See my above comment. Why does this happen in 2 places?

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs (hunk 9 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: I'm like 89% sure that that's not right.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs (hunk 18 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs]: Nope. This free happens on synchronous completion and the other happens on async completion.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 20]
```diff
@@ -101,8 +101,7 @@ private static NativeMethods.REQUEST_NOTIFICATION_STATUS OnAsyncCompletion(IntPt
         private static void CompleteRequest(IISHttpContext context)
         {
             // Post completion after completing the request to resume the state machine
-            //context.PostCompletion();
-            context.IndicateCompletion(NativeMethods.REQUEST_NOTIFICATION_STATUS.RQ_NOTIFICATION_CONTINUE);
+            context.PostCompletion(NativeMethods.REQUEST_NOTIFICATION_STATUS.RQ_NOTIFICATION_CONTINUE);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: I'm assuming this needs to change?

Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: What do you mean by that? I have another PR that will change the value of request notification status based on success or failure.

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: NM I read the rest of the codez

**Comment by jkotalik** [comment]
*cough* *cough* can you spare a green checkmark 😄

**Comment by jkotalik** [comment]
pending https://github.com/aspnet/AspNetCoreModule/pull/212

**Pull request merged**